pkg/scheduling/v2/scheduler.go

```diff
@@ -0,0 +1,602 @@
+			continue
+		}
+
+		actionId := workerActionTuple.ActionId.String
+		workerId := sqlchelpers.UUIDToStr(workerActionTuple.WorkerId)
+
+		actionsToWorkerIds[actionId] = append(actionsToWorkerIds[actionId], workerId)
+		workerIdsToActions[workerId] = append(workerIdsToActions[workerId], actionId)
+	}
+
+	// FUNCTION 1: determine which actions should be replenished. Logic is the following:
+	// - zero or one slots for an action: replenish all slots
+	// - some slots for an action: replenish if 50% of slots have been used, or have expired
+	// - more workers available for an action than previously: fully replenish
+	// - otherwise, do not replenish
+	actionsToReplenish := make(map[string]bool)
+	s.actionsMu.RLock()
+
+	for actionId, workers := range actionsToWorkerIds {
+		if mustReplenish {
+			actionsToReplenish[actionId] = true
+			continue
+		}
+
+		// if the action is not in the map, it should be replenished
+		if _, ok := s.actions[actionId]; !ok {
+			actionsToReplenish[actionId] = true
+			continue
+		}
+
+		storedAction := s.actions[actionId]
+
+		// determine if we match the conditions above
+		var replenish bool
+		activeCount := storedAction.activeCount()
+
+		if activeCount == 0 {
+			replenish = true
+		} else if activeCount <= (storedAction.lastReplenishedSlotCount / 2) {
+			replenish = true
+		} else if len(workers) > storedAction.lastReplenishedWorkerCount {
+			replenish = true
+		}
+
+		actionsToReplenish[actionId] = replenish
+	}
+
+	s.actionsMu.RUnlock()
+
+	// FUNCTION 2: for each action which should be replenished, load the available slots
+	uniqueWorkerIds := make(map[string]bool)
+
+	for actionId, replenish := range actionsToReplenish {
+		if !replenish {
+			continue
+		}
+
+		workerIds := actionsToWorkerIds[actionId]
+
+		for _, workerId := range workerIds {
+			uniqueWorkerIds[workerId] = true
+		}
+	}
+
+	workerUUIDs := make([]pgtype.UUID, 0, len(uniqueWorkerIds))
+
+	for workerId := range uniqueWorkerIds {
+		workerUUIDs = append(workerUUIDs, sqlchelpers.UUIDFromStr(workerId))
+	}
+
+	availableSlots, err := s.repo.ListAvailableSlotsForWorkers(ctx, dbsqlc.ListAvailableSlotsForWorkersParams{
+		Tenantid:  s.tenantId,
+		Workerids: workerUUIDs,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// FUNCTION 3: list unacked slots (so they're not counted towards the worker slot count)
+	workersToUnackedSlots := make(map[string][]*slot)
+
+	// TODO: CHECK CORRECTNESS OF UNACKED SLOTS, PERHAPS WE SHOULD LOCK THIS FOR THE ENTIRE REPLENISH?
+	s.unackedMu.Lock()
+	defer s.unackedMu.Unlock()
+
+	for _, unackedSlot := range s.unackedSlots {
+		s := unackedSlot
+		workerId := s.getWorkerId()
+
+		if _, ok := workersToUnackedSlots[workerId]; !ok {
+			workersToUnackedSlots[workerId] = make([]*slot, 0)
+		}
+
+		workersToUnackedSlots[workerId] = append(workersToUnackedSlots[workerId], s)
+	}
+
+	// FUNCTION 4: write the new slots to the scheduler and clean up expired slots
+	now := time.Now()
+	expires := now.Add(5 * time.Second)
```

Comment:
what is this arb expires?

Reply:
removed the expires logic as I don't think it's doing anything